bin/pycbc_make_skymap

```diff
@@ -326,17 +352,23 @@ if __name__ == '__main__':
                              'production analyses!')
     parser.add_argument('--gracedb-server', metavar='URL', default=None,
                         help='URL of GraceDB server API for uploading events. ')
-
+    parser.add_argument('--custom-frame-file', type=str, nargs='+',
+                        help='lists of local frame files, e.g., H1:/path/to/frame/file L1:/path/to/frame/file')
+    
     opt = parser.parse_args()
 
     frame_type_dict = {f.split(':')[0]: f.split(':')[1] for f in opt.frame_type} \
             if opt.frame_type is not None else None
     chan_name_dict = {f.split(':')[0]: f for f in opt.channel_name} \
             if opt.channel_name is not None else None
+    custom_frame_dict ={f.split(':')[0]: f.split(':')[1] for f in opt.custom_frame_file} \
+        if opt.custom_frame_file is not None else None
+
+    
 
     main(opt.trig_time, opt.mass1, opt.mass2,
          opt.spin1z, opt.spin2z, opt.f_low, opt.ifar, opt.ifos,
          opt.thresh_SNR, opt.ligolw_psd_output, opt.ligolw_event_output,
          frame_types=frame_type_dict, channel_names=chan_name_dict,
          gracedb_server=opt.gracedb_server,
-         test_event=not opt.enable_production_gracedb_upload)
+         test_event=not opt.enable_production_gracedb_upload, custom_frame_files=custom_frame_dict)
```

Comment:
Please put `custom_frame_files=custom_frame_dict` on a new line indented consistently with the previous arguments.

Reply:
Thank you. I have changed it.

Reply:
> @kentaromogushi I left two minor coding-style comments, otherwise this looks good. Have you tested it?

Yes, I have tested. A custom frame with a correct trigger time and a channel name gives a sky map properly. Also, it works without a custom frame file, for example, for S190814bv, which is used as a test for the original Fergus's code. To test if the code correctly quits due to inappropriate custom frame files, I have conducted several tests below.   

Here are the tests. The frame file I used for the test has a BBH injection with a merger time at GPS of 1235434218. The length of this frame file is 500 sec evenly around the merger time.  The channel name is H1:GDS-CALIB_STRAIN. 
 
-----------------------------------------------------------------------------+
      Set the trigger time in the argument to be the merger time - 200 sec         |
     --> properly quit it because the start time of the frame is not long enough |
           for pycbc_single_template                                                                             |
-----------------------------------------------------------------------------+
(py36_rapid_sky) [kentaro.mogushi@ldas-pcdev6 pycbc_make_skymap_test]$ ./pycbc_make_skymap --mass1 35 --mass2 29 --spin1z 0 --spin2z 0 --ifos H1 --custom-frame-file H1:raw_50_H1.gwf --trig-time 1235434018
2020-06-12 22:15:07,018 check if the segment in the custom frame file is safe
2020-06-12 22:15:08,203 Start time of 1235433968 should be before the required start time 1235433826
2020-06-12 22:15:08,203 Exit the program

-----------------------------------------------------------------------------+
      Set the trigger time in the argument to be the merger time + 200 sec         |
     --> properly quit it because the end time of the frame is not long enough   |
           for pycbc_single_template                                                                            |
-----------------------------------------------------------------------------+
(py36_rapid_sky) [kentaro.mogushi@ldas-pcdev6 pycbc_make_skymap_test]$ ./pycbc_make_skymap --mass1 35 --mass2 29 --spin1z 0 --spin2z 0 --ifos H1 --custom-frame-file H1:raw_50_H1.gwf --trig-time 1235434418
2020-06-12 22:19:44,051 check if the segment in the custom frame file is safe
2020-06-12 22:19:45,425 End time of 1235434468 should be after the required end time 1235434482
2020-06-12 22:19:45,425 Exit the program

----------------------------------------+
   choose the wrong channel name          |
   --> properly quit it                                  |
----------------------------------------+
(py36_rapid_sky) [kentaro.mogushi@ldas-pcdev6 pycbc_make_skymap_test]$ ./pycbc_make_skymap --mass1 35 --mass2 29 --spin1z 0 --spin2z 0 --ifos H1 --custom-frame-file H1:raw_50_H1.gwf --trig-time 1235434218 --channel-name H1:wrong
2020-06-12 22:21:32,941 check if the segment in the custom frame file is safe
XLAL Error - XLALFrameUFrChanRead_FrameL_ (LALFrameL.c:880): Channel H1:wrong not found
XLAL Error - XLALFrameUFrChanRead_FrameL_ (LALFrameL.c:880): Wrong name
XLAL Error - XLALFrFileQueryChanVectorLength (LALFrameIO.c:199): Wrong name
2020-06-12 22:21:32,942 Channel name in raw_50_H1.gwf is not H1:wrong
2020-06-12 22:21:32,943 Unable to open it
2020-06-12 22:21:32,943 Exit the program

 -------------------------------------------------------------------------------------+
  choose a trigger time such that the automatically chosen channel name is wrong       |
  ,i.e., choose a trigger time in O2                                                                                          |
  properly quit it                                                                                                                       |
--------------------------------------------------------------------------------------+
(py36_rapid_sky) [kentaro.mogushi@ldas-pcdev6 pycbc_make_skymap_test]$ ./pycbc_make_skymap --mass1 35 --mass2 29 --spin1z 0 --spin2z 0 --ifos H1 --custom-frame-file H1:raw_50_H1.gwf --trig-time 1235433518
2020-06-12 22:24:32,370 check if the segment in the custom frame file is safe
XLAL Error - XLALFrameUFrChanRead_FrameL_ (LALFrameL.c:880): Channel H1:DCH-CLEAN_STRAIN_C02 not found
XLAL Error - XLALFrameUFrChanRead_FrameL_ (LALFrameL.c:880): Wrong name
XLAL Error - XLALFrFileQueryChanVectorLength (LALFrameIO.c:199): Wrong name
2020-06-12 22:24:32,377 Channel name in raw_50_H1.gwf is not H1:DCH-CLEAN_STRAIN_C02
2020-06-12 22:24:32,377 Unable to open it
2020-06-12 22:24:32,377 Exit the program

 





----

Reply:
Yes, I have tested. A custom frame with a correct trigger time and a channel name gives a sky map properly. Also, it works without a custom frame file, for example, for S190814bv, which is used as a test for the original Fergus's code. To test if the code correctly quits due to inappropriate custom frame files, I have conducted several tests below.   

Here are the tests. The frame file I used for the test has a BBH injection with a merger time at GPS of 1235434218. The length of this frame file is 500 sec evenly around the merger time.  The channel name is H1:GDS-CALIB_STRAIN. 
 
-----------------------------------------------------------------------------+
      Set the trigger time in the argument to be the merger time - 200 sec         |
     --> properly quit it because the start time of the frame is not long enough |
           for pycbc_single_template                                                                             |
-----------------------------------------------------------------------------+
(py36_rapid_sky) [kentaro.mogushi@ldas-pcdev6 pycbc_make_skymap_test]$ ./pycbc_make_skymap --mass1 35 --mass2 29 --spin1z 0 --spin2z 0 --ifos H1 --custom-frame-file H1:raw_50_H1.gwf --trig-time 1235434018
2020-06-12 22:15:07,018 check if the segment in the custom frame file is safe
2020-06-12 22:15:08,203 Start time of 1235433968 should be before the required start time 1235433826
2020-06-12 22:15:08,203 Exit the program

-----------------------------------------------------------------------------+
      Set the trigger time in the argument to be the merger time + 200 sec         |
     --> properly quit it because the end time of the frame is not long enough   |
           for pycbc_single_template                                                                            |
-----------------------------------------------------------------------------+
(py36_rapid_sky) [kentaro.mogushi@ldas-pcdev6 pycbc_make_skymap_test]$ ./pycbc_make_skymap --mass1 35 --mass2 29 --spin1z 0 --spin2z 0 --ifos H1 --custom-frame-file H1:raw_50_H1.gwf --trig-time 1235434418
2020-06-12 22:19:44,051 check if the segment in the custom frame file is safe
2020-06-12 22:19:45,425 End time of 1235434468 should be after the required end time 1235434482
2020-06-12 22:19:45,425 Exit the program

----------------------------------------+
   choose the wrong channel name          |
   --> properly quit it                                  |
----------------------------------------+
(py36_rapid_sky) [kentaro.mogushi@ldas-pcdev6 pycbc_make_skymap_test]$ ./pycbc_make_skymap --mass1 35 --mass2 29 --spin1z 0 --spin2z 0 --ifos H1 --custom-frame-file H1:raw_50_H1.gwf --trig-time 1235434218 --channel-name H1:wrong
2020-06-12 22:21:32,941 check if the segment in the custom frame file is safe
XLAL Error - XLALFrameUFrChanRead_FrameL_ (LALFrameL.c:880): Channel H1:wrong not found
XLAL Error - XLALFrameUFrChanRead_FrameL_ (LALFrameL.c:880): Wrong name
XLAL Error - XLALFrFileQueryChanVectorLength (LALFrameIO.c:199): Wrong name
2020-06-12 22:21:32,942 Channel name in raw_50_H1.gwf is not H1:wrong
2020-06-12 22:21:32,943 Unable to open it
2020-06-12 22:21:32,943 Exit the program

 -------------------------------------------------------------------------------------+
  choose a trigger time such that the automatically chosen channel name is wrong       |
  ,i.e., choose a trigger time in O2                                                                                          |
  properly quit it                                                                                                                       |
--------------------------------------------------------------------------------------+
(py36_rapid_sky) [kentaro.mogushi@ldas-pcdev6 pycbc_make_skymap_test]$ ./pycbc_make_skymap --mass1 35 --mass2 29 --spin1z 0 --spin2z 0 --ifos H1 --custom-frame-file H1:raw_50_H1.gwf --trig-time 1235433518
2020-06-12 22:24:32,370 check if the segment in the custom frame file is safe
XLAL Error - XLALFrameUFrChanRead_FrameL_ (LALFrameL.c:880): Channel H1:DCH-CLEAN_STRAIN_C02 not found
XLAL Error - XLALFrameUFrChanRead_FrameL_ (LALFrameL.c:880): Wrong name
XLAL Error - XLALFrFileQueryChanVectorLength (LALFrameIO.c:199): Wrong name
2020-06-12 22:24:32,377 Channel name in raw_50_H1.gwf is not H1:DCH-CLEAN_STRAIN_C02
2020-06-12 22:24:32,377 Unable to open it
2020-06-12 22:24:32,377 Exit the program